### PR TITLE
Ner route traffic aware

### PR DESCRIPTION
**Pull request by rowleya**
This adds two algorithms to the set available:
 - NerRouteTrafficAware.  This is the same as NerRoute except that, when the directions to be taken in a route are being considered, instead of taking the longest dimension first, this looks at the traffic that has been assigned to each of the combinations and chooses the one with the least traffic on it.
 - GlobalZonedRoutingInfoAllocator.  Like ZonedRoutingInfoAllocator, this allocates keys by ApplicationVertex.  The difference is that the mask is the same for every vertex.  It doesn't work with constrained keys currently (but that is fine for what it is useful for).

### CODE REVIEW

**Comment by Christian-B** [comment]
Ok to merge if https://github.com/SpiNNakerManchester/sPyNNaker8/pull/417 goes Green

File: pacman/operations/algorithms_metadata.xml [hunk 1]
```diff
@@ -385,6 +385,43 @@
...
+    <algorithm name="GlobalZonedRoutingInfoAllocator">
+        <python_module>pacman.operations.routing_info_allocator_algorithms.global_zoned_routing_info_allocator</python_module>
+        <python_class>GlobalZonedRoutingInfoAllocator</python_class>
+        <input_definitions>
+            <parameter>
+                <param_name>application_graph</param_name>
+                <param_type>MemoryApplicationGraph</param_type>
+            </parameter>
+            <parameter>
+                <param_name>graph_mapper</param_name>
```
Comment (alan-stokes) [inline pacman/operations/algorithms_metadata.xml:397]: no longer exists

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+from pacman.model.routing_info import (
+    RoutingInfo, PartitionRoutingInfo, BaseKeyAndMask)
+from pacman.utilities.utility_calls import (
+    check_algorithm_can_support_constraints)
+from pacman.exceptions import PacmanRouteInfoAllocationException
+from pacman.model.constraints.key_allocator_constraints import (
+    AbstractKeyAllocatorConstraint, ContiguousKeyRangeContraint)
+from pacman.model.graphs.common import EdgeTrafficType
+
+KEY_SIZE = 32
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:28]: pacman constants has this. no need to clone the constant

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    .. note::
+        No constraints are supported, and that the number of keys\
+        required by each edge must be 2048 or less, and that all edges coming\
+        out of a vertex will be given the same key/mask assignment.
+    """
+
+    __slots__ = [
+        # Passed in parameters
+        "_application_graph",
+        "_graph_mapper",
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:44]: nope

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        "_machine_graph",
+        "_placements",
+        "_n_keys_map",
+        "_n_bits_partition",
+        "_n_bits_machine",
+        "_n_bits_atoms"
+    ]
+    # pylint: disable=attribute-defined-outside-init
+
+    def __call__(self, application_graph, graph_mapper, machine_graph,
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:54]: still nope on the graph mapper

Comment (alan-stokes) [reply pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:54]: (cwl) its not even doced in the call docs. so least you dont need to fix that. but still. bad on you for not docing the dead param. lol

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        :rtype:\
+            :py:class:`pacman.model.routing_info.PartitionRoutingInfo`
+        :raise pacman.exceptions.PacmanRouteInfoAllocationException: \
+            If something goes wrong with the allocation
+        """
+
+        # check that this algorithm supports the constraints put onto the
+        # partitions
+        self._application_graph = application_graph
+        self._graph_mapper = graph_mapper
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:79]: nope

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        progress = ProgressBar(
+            self._application_graph.n_vertices, "Calculating zones")
+        max_partitions = 0
+        max_machine_vertices = 0
+        max_keys = 0
+        n_app_vertices = 0
+
+        for app_vertex in progress.over(self._application_graph.vertices):
+            app_max_partitions = 0
+            machine_vertices = self._graph_mapper.get_machine_vertices(
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:102]: nope

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+            self._application_graph.n_vertices, "Allocating routing keys")
+        routing_infos = RoutingInfo()
+        by_app_vertex = dict()
+        mask = 0xFFFFFFFF - ((2 ** self._n_bits_atoms) - 1)
+        app_bits = (
+            self._n_bits_atoms + self._n_bits_machine + self._n_bits_partition)
+        app_mask = 0xFFFFFFFF - ((2 ** app_bits) - 1)
+        app_index = 0
+        for app_vertex in progress.over(self._application_graph.vertices):
+            machine_vertices = self._graph_mapper.get_machine_vertices(
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:152]: nope

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+                app_key = app_index << (self._n_bits_machine +
+                                        self._n_bits_partition +
+                                        self._n_bits_atoms)
+                by_app_vertex[app_vertex] = BaseKeyAndMask(
+                    base_key=app_key, mask=app_mask)
+                app_index += 1
+
+        return routing_infos, by_app_vertex
+
+    def _bites_needed(self, size):
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:183]: definte static method here.

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+                partitions = self._machine_graph.\
+                    get_outgoing_edge_partitions_starting_at_vertex(vertex)
+                part_index = 0
+                for partition in partitions:
+                    if partition.traffic_type == EdgeTrafficType.MULTICAST:
+                        key = app_index
+                        key = (key << self._n_bits_machine) | machine_index
+                        key = (key << self._n_bits_partition) | part_index
+                        key = key << self._n_bits_atoms
+                        key_and_mask = BaseKeyAndMask(
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:165]: can be 1 line. not 2

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+                            base_key=key, mask=mask)
+                        info = PartitionRoutingInfo(
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:167]: same. 1 liner

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+                            [key_and_mask], partition)
+                        routing_infos.add_partition_info(info)
+                        part_index += 1
+                if part_index > 0:
+                    machine_index += 1
+            if machine_index > 0:
+                app_key = app_index << (self._n_bits_machine +
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:174]: yuck. im sure that can be made to read nicer

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+                    self._n_bits_machine, self._n_bits_partition,
+                    self._n_bits_atoms, n_bits_vertices))
+        return max_partitions
+
+    def _simple_allocate(self):
+        progress = ProgressBar(
+            self._application_graph.n_vertices, "Allocating routing keys")
+        routing_infos = RoutingInfo()
+        by_app_vertex = dict()
+        mask = 0xFFFFFFFF - ((2 ** self._n_bits_atoms) - 1)
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:149]: magic number. used twice. named constant?

Comment (alan-stokes) [reply pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:149]: also why isnt that in pacman constants. you'd assume we use that elsewhere............

Comment (alan-stokes) [reply pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:149]: oh its in pacman constants! USE THE NAMED CONSTANT!

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py (hunk 8 above)
Comment (alan-stokes) [reply pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py]: spelling problem here. if you want to not have it static. named constant for the 2? else, it should go to some helpful functions somewhere.

Comment (rowleya) [reply pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py]: Think you are looking at outdated code...

Regarding the 2, this is the binary base power.  It should be well understood without needing to be named.

**Comment by rowleya** [comment]
I believe that all changes have now been made.  I don't think that naming "2" as the binary base is needed surely...

File: pacman/operations/router_algorithms/ner_route.py [hunk 13]
```diff
@@ -83,7 +84,7 @@ def _convert_a_route(
             routing_tables, partition, None, next_incoming_link, next_hop)
```
Comment (alan-stokes) [inline pacman/operations/router_algorithms/ner_route.py:84]: cant have None there. the docs for _convert_a_route on lines 54 and 55 say it must be a int.

File: pacman/operations/router_algorithms/ner_route.py [hunk 14]
```diff
@@ -83,7 +84,7 @@ def _convert_a_route(
...
 
 
-def _ner_net(source, destinations, machine):
+def _ner_net(source, destinations, machine, vector_to_nodes):
```
Comment (alan-stokes) [inline pacman/operations/router_algorithms/ner_route.py:87]: missing doc for vector_to_nodes

File: pacman/operations/router_algorithms/ner_route.py [hunk 15]
```diff
@@ -458,14 +460,15 @@ def _do_route(source_vertex, post_vertexes, machine, placements):
     :param iterable(MachineVertex) post_vertexes:
```
Comment (alan-stokes) [inline pacman/operations/router_algorithms/ner_route.py:460]: none of these docs filled in.

File: pacman/operations/router_algorithms/ner_route.py [hunk 16]
```diff
@@ -585,6 +624,40 @@ def _longest_dimension_first(vector, start, machine):
...
+
+    for source_vertex in progress_bar.over(machine_graph.vertices):
+        # handle the vertex edges
+        for partition in machine_graph.\
+                get_outgoing_edge_partitions_starting_at_vertex(
+                    source_vertex):
+            if partition.traffic_type == EdgeTrafficType.MULTICAST:
+                post_vertexes = list(
+                    e.post_vertex for e in partition.edges)
+                routingtree = _do_route(
```
Comment (alan-stokes) [inline pacman/operations/router_algorithms/ner_route.py:648]: routingtree vs routing_tree?

File: pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+        """ Allocate the keys given the calculations done previously
+        """
+        progress = ProgressBar(
+            self.__application_graph.n_vertices, "Allocating routing keys")
+        routing_infos = RoutingInfo()
+
+        # Keep track of the application vertex keys and masks
+        by_app_vertex = dict()
+
+        # Work out the mask to use in the key and the applcation-vertex key
```
Comment (alan-stokes) [inline pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py:201]: application vs applcation

**Comment by alan-stokes** [comment]
went back to find my "outdated" static method. annoyingly whilst i went there, pycharm flagged these up to me. i admit they not directly in the PR. but def should be looked at, and as your already here...............

oh and on that static method. you win coz you moved it outside the class. Pycharm cant moan at that. lol. 

line 65 on ner_route.py link = None but then on line 73 if link is not None:. but nothing changes link inbetween, so that code CAN NEVER BE RAN!

line 171 of ner_route.py redundant brackets on return. 

same issue on lines 188 and 190, 266, 507, 515,

line 254. direction can be None, but the docs on _is_linked says it needs to be a int. 

line 419 meerly and ocurred. i think ocurred should be occurred. meerly dont know if thats just pycharm being funny.

thats it folks. I think after all these, i couldn't argue even if i wanted to.

File: pacman/operations/router_algorithms/ner_route.py (hunk 13 above)
Comment (rowleya) [reply pacman/operations/router_algorithms/ner_route.py:84]: If you want to do a tidy up of existing code, please feel free.

Comment (alan-stokes) [reply pacman/operations/router_algorithms/ner_route.py:80]: ..................... pooy to you then. lol.

**Pull request merged**